### ISSUE TITLE
OpenSC testing on Windows 

### ISSUE THREAD

**Issue by dengert** [OpenSC/OpenSC#586]
@viktorTarasov @frankmorgner @LudovicRousseau @JenkinsForOpenSC 

The effort needed for an end user or even a developer to test Pull Requests for Windows is 
not working. An end user or even a developer such as myself is not willing to install all the development tools on Windows, but could install an MSI file if it was available. 

We keep changing how to build OpenSC for Windows and what  continues build environment is used. Jenkins used to work. #585 says the AppVoyer MSIs are missing files. 

http://sourceforge.net/projects/opensc/files/OpenSC/nightly/
for Windows are a month old. 

If we wanted to be serious about Windows we need to make the testing eaiser.

**Comment by frankmorgner** [comment]
could you please check with `depends.exe` (http://www.dependencywalker.com/) which files are missing?

**Comment by frankmorgner** [comment]
maybe some redistributables are missing https://msdn.microsoft.com/en-us/library/vstudio/8kche8ah%28v=vs.100%29.aspx

**Comment by frankmorgner** [comment]
`msvcr100.dll` is needed with `/MT` that could be it (https://github.com/OpenSC/OpenSC/blob/master/win32/Make.rules.mak#L86). However, I think this was never included in OpenSC, so I'm wondering if this library was linked statically previously...

**Comment by dengert** [comment]
This should be sent to #585

On 10/17/2015 10:32 AM, Frank Morgner wrote:
> could you please check with |depends.exe| (http://www.dependencywalker.com/) which files are missing?
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/586#issuecomment-148925115>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by frankmorgner** [comment]
@imasonaz could you check with `depends.exe` which libraries are missing?

**Comment by dengert** [comment]
#585 reported problems with missing files.

I was hoping that this thread #586 would be a more general review of how OpenSC allows users and developers
to test PRs and what OpenSC can do to get this working using the continuous build processes, or the process used
to build the actuall releases.


On 10/17/2015 10:38 AM, Frank Morgner wrote:
> maybe some redistributables are missing https://msdn.microsoft.com/en-us/library/vstudio/8kche8ah%28v=vs.100%29.aspx
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/586#issuecomment-148925401>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by frankmorgner** [comment]
How can this issue be solved, what are your expectations?

**Comment by dengert** [comment]
My expectations are:

A user can download win32 and win64 MSIs based on a PR or nightly builds.
The MSIs should be built using the same method used for the official release.

A user can then test a PR or nightly builds without having to build it.
This will allow many users to test patches on Windows without a windows development environment.

The WIKI needs to instruct the user on how to find the MSIs.

We appear to be close. Failure of the build machines/process appears to be the biggest problem.
MSIs of the nightly builds would be a big improvement.


On 10/17/2015 12:56 PM, Frank Morgner wrote:
> How can this issue be solved, what are your expectations?
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/586#issuecomment-148938101>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by viktorTarasov** [comment]
Normally buid of PRs is restored.
Currently I have no active project with OpenSC and do not monitor closely the CI environment.

**Comment by vletoux** [comment]
Sorry but I do not understand the problem because I can download the msi file from my PR.
Maybe I missed something

See the following screenshots made from my PR:

![build1](https://cloud.githubusercontent.com/assets/10632326/10620882/c7531026-777d-11e5-88de-e101c16179b5.png)
==
![build2](https://cloud.githubusercontent.com/assets/10632326/10620886/ca57ab6a-777d-11e5-83df-c270e2520261.png)
==
![build3](https://cloud.githubusercontent.com/assets/10632326/10620891/cc87eb52-777d-11e5-972f-47cfd56d5e58.png)

**Comment by imasonaz** [comment]
@vletoux-
Read #585, when installing some of the latest nightly builds, some .dlls
are missing/broken. I can provide specifics, but it is reproducible by
uninstalling OpenSC, and installing a nightly build from Source Forge or
the other site (name escapes me at the moment).

**Comment by vletoux** [comment]
The library missing in the setup is libeay32.dll

![depends](https://cloud.githubusercontent.com/assets/10632326/10629698/6a8f3b9a-77d1-11e5-973c-de29ef749dfe.png)
![depends2](https://cloud.githubusercontent.com/assets/10632326/10629700/6ec169ea-77d1-11e5-8a17-06ff5551fafd.png)

**Comment by vletoux** [comment]
@viktorTarasov the win32 is failing because the file wcautil.h cannot be located.
It is part of the WIX SDK
The WIX SDK path is defined in win32/Make.rules.mak
The path which were used in your build machine was: "C:\Program Files (x86)\Windows Installer XML v3.6"
=> there is a conflict between AppVeyor (v3.10) and your build machine (v3.60)

Upgrading WIX from 3.6 to 3.10 in the build machine should solve the problem

**Comment by viktorTarasov** [comment]
On 10/21/2015 11:43 AM, vletoux wrote:
>
> @viktorTarasov <https://github.com/viktorTarasov> the win32 is failing because the file wcautil.h cannot be located.
> It is part of the WIX SDK
> The WIX SDK path is defined in win32/Make.rules.mak
> The path which were used in your build machine was: "C:\Program Files (x86)\Windows Installer XML v3.6"
> => there is a conflict between AppVeyor (v3.10) and your build machine (v3.60)
>
> Upgrading WIX from 3.6 to 3.10 in the build machine should solve the problem
>

Yes, I see it,
the question is : do we need the 'default' Windows builds (that uses private VMs with CI slaves)
when AppVeyor do the same jobs (on public platforms) ?
Just to add the sending of MSI artifacts to sourceforge and it will completely replace the current 'default' Windows builds.

**Comment by frankmorgner** [comment]
@viktorTarasov The question is whether you're trusting AppVeyor to produce the correct binaries. I would prefer to get my binaries from the OpenSC team even if it uses the same build chain as AppVeyor...

**Comment by vletoux** [comment]
I've run the setup of the "full version" produced by AppVeyor on a VM without any component installed except OpenSC.
The minidriver / pkcs11 module can't be loaded because the OpenSSL dll (libeay32.dll, ...) can't be found.

Cause: the wix file
the libeay32.dll is installed, but in the tools directory and the minidriver/pkcs11 library is installed in c:\windows\system32
=> the dll is not found

This can be fixed:
- solution 1)
Install the minidriver and the pkcs11 library in the at the root of the directory in the program files folder and put a second copy of the OpenSSL librairies
No problem for the minidriver registration because i'm using it for my company for several years.
But what if the applications loading the pkcs11 library do not support using a path with a space (program_[space]_files)

- solution 2)
install the OpenSSL library in system 32 (and what if there is conflict with other installations ?)

- solution 3)
Add the tools directory to the path environment variable. OpenSSL lib will be found and better, the tools will be available directly from the command line

- solution 4)
Implement a hook to load the library stored in the tool directory if the dll cannot be found
(requiring "delay load library" loading + __pfnDliNotifyHook2)

**Comment by dengert** [comment]
According to depend 2.2.6000 64 bit, the c:\Windows\system32\opensc-minidriver.dll (64 bit module)

is trying to load LIBEAY32.DLL as a x86 "Error: Modules with different CPU types were found."

On my system it is finding LIBEAY32.DLL at:
C:\program files (x86)\intel\icls client
https://communities.intel.com/thread/46711
talks about Intels problem this problem.

solution 1) may work...
solution 2) may cause conflicts
solution 3) I don't think will work, when minidriver is run from LSA, I don't believe there is a user or a path.


What about
  LoadLibraryEx function with LOAD_WITH_ALTERED_SEARCH_PATH
The search path could be in the registry under OpenSC project.

See:
https://msdn.microsoft.com/en-us/library/windows/desktop/ms682586(v=vs.85).aspx



On 10/25/2015 3:17 AM, vletoux wrote:
> I've run the setup of the "full version" produced by AppVeyor on a VM without any component installed except OpenSC.
> The minidriver / pkcs11 module can't be loaded because the OpenSSL dll (libeay32.dll, ...) can't be found.
>
> Cause: the wix file
> the libeay32.dll is installed, but in the tools directory and the minidriver/pkcs11 library is installed in c:\windows\system32
> => the dll is not found
>
> This can be fixed:
>
>   *
>
>     solution 1)
>     Install the minidriver and the pkcs11 library in the at the root of the directory in the program files folder and put a second copy of the OpenSSL librairies
>     No problem for the minidriver registration because i'm using it for my company for several years.
>     But what if the applications loading the pkcs11 library do not support using a path with a space (program_[space]_files)
>
>   *
>
>     solution 2)
>     install the OpenSSL library in system 32 (and what if there is conflict with other installations ?)
>
>   *
>
>     solution 3)
>     Add the tools directory to the path environment variable. OpenSSL lib will be found and better, the tools will be available directly from the command line
>
>   *
>
>     solution 4)
>     Implement a hook to load the library stored in the tool directory if the dll cannot be found
>     (requiring "delay load library" loading + __pfnDliNotifyHook2)
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/586#issuecomment-150900118>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by vletoux** [comment]
This message means that the system found the dll but not a version compatible with the processor architecture.
This typically related to x86 software registered on the path variable of a x64 system.

Any opinion/vote in favor of one of the solution I proposed?

**Comment by frankmorgner** [comment]
We used to link statically against openssl and zlib which I did not get working anymore, see 428385d8b44888482d1e3d066d930258194a5285 for a detailed explanation. Maybe you can have a deeper look at this solution (solution 5)?

If that is not successful I think that modifying PATH (solution 3) is the easiest solution. If other programs in the PATH are bringing their own version of OpenSSL this may give a conflict. Solution 4 looks more robust; would it be complicated to do?

**Comment by dengert** [comment]
No comments on: LoadLibraryEx function with LOAD_WITH_ALTERED_SEARCH_PATH ?

What about side by side assembles? winsxs

How about renaming the libs. (Issue is finding them.)


I still believe using the PATH will not work when minidriver is run from login.
(As a test, adding a getenv to minidriver then writing to the md.log would show this.)
Also as pointed out above too many other applications are using different versions of libeay.dll,
and using PATH.

Does this help?
https://msdn.microsoft.com/en-us/library/windows/desktop/ms682586(v=vs.85).aspx

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by vletoux** [comment]
Do you mean by the solution 5 the comment @dengert did, but which was removed from the github comment ?
Using a LoadLibrary flag is not possible because the library is linked and not loaded by LoadLibrary.
What I called the solution 4 is the closest solution to that.
We transform the dll as a delay loaded dll and we have a hook to do the same thing that the one suggested to LoadLibrary.
I used this trick to be able to use Windows 7 native function which are not available on Windows XP while not breaking the compatibility with Windows XP.

Solution 4 is not as complicated as it seems because I've already done it.
(see http://sourceforge.net/p/eidauthenticate/code/HEAD/tree/EIDCardLibrary/XPCompatibility.cpp line 329 to handle dliFailLoadLib)

I looked also SXS which can be done, but only on .exe side, not the dll itself & starting only with Windows 7 (probe path).

The path variable can be used, because you have one defined at the user level, and one at the machine level. Only the machine level will make the SC logon work.

**Comment by vletoux** [comment]
@frankmorgner 
Some stupid question regarding 428385d
Have you tried /Gs or adding the missing report function (http://cubicspot.blogspot.fr/2015/07/solving-unresolved-external-symbol.html) ?

**Comment by dengert** [comment]
Solution 4 is sounding better, with the location of the libs stored in the registry under HKEY_LOCAL_MACHINE\SOFTWARE\OpenSC Project and HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\OpenSC Project

Using PATH will just cause problems, as will env variables.

**Comment by vletoux** [comment]
@dengert Maybe you were right: AddDllDirectory can be coupled with Dll delay load.or you can do manually the loadlibrary call with the delay loading notification hook.
Nut actually, I'm more in favor of fixing the linkage problem with static library (adding the /GS or providing a definition for the missing function with maybe a #define for older version of VS)

**Comment by dengert** [comment]
Additional question:

On dual installs, with 64 and 32 bits versions, why do we have two opensc.conf  and two sets of profiles. Could  there be just one under c:\Program Data\ ?

**Comment by frankmorgner** [comment]
@vletoux no, I didn't try that.

**Comment by vletoux** [comment]
@dengert you can even make a x64 containing both x86 and x64 binairies with a single setup. You just need in the build machine to build both projects. Adjusting the setup is not a problem with me ;-)

@frankmorgner Would you like to try again to link statically OpenSSL if I give a hand to you ?
I'm rusty on linux / unix but I know now pretty well all Windows trick & issues.

**Comment by frankmorgner** [comment]
For me, this would only be trial and error. I think you will solve this better and faster than me! (Which isn't very surprising since I did *not* solve this at all :wink:)

**Comment by vletoux** [comment]
yes but I didn't now the library you tried to use nor how to manage the build machine.
Just redo the openssl .lib with /GS, and I'll check the result and provide advices

**Comment by frankmorgner** [comment]
My setup and build procedure is exactly as described in AppVeyor, see https://github.com/OpenSC/OpenSC/blob/master/appveyor.yml. I don't think we should build the dependencies ourselves. AppVeyor uses more or less official binaries of zlib and openssl, so do I.

**Comment by vletoux** [comment]
Dependancy with OpenSSL fixed in #597

**Comment by vletoux** [comment]
PR #600 compliant with @frankmorgner comments (#597 closed)

**Comment by frankmorgner** [comment]
I think there is the same problem with zlib, that gets loaded as shared library...

**Comment by vletoux** [comment]
Checked all .exe / .dll with depends without having found a zlib library needs

![depends](https://cloud.githubusercontent.com/assets/10632326/10870844/885c638e-80d5-11e5-9f3f-5cf701060ebb.png)

**Comment by frankmorgner** [comment]
@vletoux could it be that this is a result of delayed library loading? zlib is only used by piv, dnie, infocamere and actalis to decompress a certificate...

**Comment by dengert** [comment]
Is the Windows build process loading zlib as static or is the build process not enabling zlib?

pkcs15-actalis,c card-dnie.c, card-piv.c, plcs15-infocamere.c and compression.c all test for ENABLE_ZLIB.

**Comment by vletoux** [comment]
@frankmorgner  a delay load library will show a hour glass like for bcrypt.dll
=> static or not enabled

**Comment by dengert** [comment]
https://ci.appveyor.com/project/LudovicRousseau/opensc/build/0.15.0.292/job/p5lw7selmi7vsldq

The AppVeyor log for VSVER=12, for x86 Release, it has for the non Light
# the prebuilt zlib is 32 bit only
and 
$env:NMAKE_EXTRA="ZLIB_DEF=/DENABLE_ZLIB ${env:NMAKE_EXTRA}"

What do we plan to do for an official

**Comment by dengert** [comment]
AppVoyer is failing again, making it impossible for those of us without a Windows build environment to test windows 64 bit. zlib for 64 bit is still needed. 

See #605 and #608

**Comment by vletoux** [comment]
With my smart phone I can see that vs2010 built is still working. That could means that appveyor upgraded to wix 3.10.1 and that the includes for vs2012 are not here anymore. (2010 is still here)
Proposition: replace 2012 by 2010 in the configuration 

For zlib we just need to locate a binary lib to include it.

**Comment by vletoux** [comment]
Just installed wix 3.10 at this moment.
There is one vs2012 directory (its absence causes the vs2012 build failures) but no vs2010 directory. But the vs2010 directory seems to be present on the build machine, else the vs10 build would have failed !

![wix3 10](https://cloud.githubusercontent.com/assets/10632326/11169676/8cb8f8e0-8bbe-11e5-8c42-57651683d997.png)

I found another way to get the zlib libraries: the file https://github.com/imazen/freeimage/blob/master/appveyor.yml describe how it is done for another projects.
Basically, the command nuget install zlib-msvc14-x64-master -Version 1.2.8.4 -Source https://ci.appveyor.com/nuget/zlib-shallow-x0n34b3pfq1b give the missing zlib library

**Comment by dengert** [comment]
Thanks! Will give it a try.

On 11/15/2015 10:34 AM, vletoux wrote:
> Just installed wix 3.10 at this moment.
> There is one vs2012 directory (its absence causes the vs2012 build failures) but no vs2010 directory. But the vs2010 directory seems to be present on the build machine, else the vs10 build would have
> failed !
>
> wix3 10 <https://cloud.githubusercontent.com/assets/10632326/11169676/8cb8f8e0-8bbe-11e5-8c42-57651683d997.png>
>
> I found another way to get the zlib libraries: the file https://github.com/imazen/freeimage/blob/master/appveyor.yml describe how it is done for another projects.
> Basically, the command nuget install zlib-msvc14-x64-master -Version 1.2.8.4 -Source https://ci.appveyor.com/nuget/zlib-shallow-x0n34b3pfq1b give the missing zlib library
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/586#issuecomment-156827726>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by vletoux** [comment]
I propose something: upload the Windows zlib / openssl static library for x86 and x64 in a dedicated repository in github/Opensc.
This way, we just have to say to Appveyor to download them.
As a bonus, we can remove the zlib dll in the Windows installation.

If you are ok, can you create a new repository ?
I'll make 2 pull requests: one to upload the .lib file in this new repository, then update the appveyor file to fix the compilation problems and modify the wix file to remove the installation of zlib.dll

**Comment by frankmorgner** [comment]
I don't like the idea of shipping unofficial binaries of other libs. It makes the build process less transparent. Could you instead try to include the build process of zlib in appveyor.yml?

Am 23. Dezember 2015 10:06:10 MEZ, schrieb vletoux <notifications@github.com>:
>I propose something: upload the Windows zlib / openssl static library
>for x86 and x64 in a dedicated repository in github/Opensc.
>This way, we just have to say to Appveyor to download them.
>As a bonus, we can remove the zlib dll in the Windows installation.
>
>If you are ok, can you create a new repository ?
>I'll make 2 pull requests: one to upload the .lib file in this new
>repository, then update the appveyor file to fix the compilation
>problems and modify the wix file to remove the installation of zlib.dll
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/OpenSC/OpenSC/issues/586#issuecomment-166838576

-- 
Frank Morgner

**Comment by vletoux** [comment]
Appveyor current compilation problems fixed in #644 
I propose to drop the VS2010 build as the x64 is known to fail and that it multiply the appveyor compilation time by 2.
=> OK ?

Will investigate if I can add the zlib build process.

I need the OpenSSL module compiled for Windows and I found out that Appveyor is not configure for both engine_pkcs11 and libp11. I can work on it if @LudovicRousseau enable the Appveyor compilation for these projects and grant me the right to do so.
=> OK ?

**Comment by frankmorgner** [comment]
Well, even without VS2010 the build will take over 20 Minutes. You don't expect immediate results from AppVeyor. The VS2010 x64 builds are already allowed to fail, an error will be ignored in the overall result of the build. I'd like to keep the status quo.

https://github.com/OpenSC/libp11 and https://github.com/OpenSC/engine_pkcs11 are not tied to https://github.com/OpenSC/OpenSC, they are (and should be) treated as separate projects. I think @nmav is doing most stuff in those projects today. You may as usual make a PR on those projects while for testing the build procedure by enabling AppVeyor on your remote.

**Comment by vletoux** [comment]
pull request #46 for libp11 to enable Appveyor compilation (artifact = libp11.dll & libp11.lib)

**Comment by vletoux** [comment]
pull request #32 for engine_pkcs11 (with rebuild of libp11) to enable Appveyor compilation (artifacts = libp11.dll, libp11.lib, engine_pkcs11.dll, engine_pkcs11.lib)

**Comment by frankmorgner** [comment]
Here are the correct links: https://github.com/OpenSC/libp11/pull/46 and https://github.com/OpenSC/engine_pkcs11/pull/32. I commented on https://github.com/OpenSC/engine_pkcs11/pull/32

**Comment by vletoux** [comment]
Last problem fixed in #645. Zlib is now included into the build process (meaning available for x64) and compiled statically. There is no need to deploy zlib1.dll anymore.
=> this issue can be closed
